batch2/batch/semaphore.py

```diff
@@ -23,24 +25,34 @@ def __init__(self, sem, weight):
         await self.sem.acquire(self.weight)
 
     async def __aexit__(self, exc_type, exc, tb):
-        await self.sem.release(self.weight)
+        self.sem.release(self.weight)
 
 
-class WeightedSemaphore:
+class FIFOWeightedSemaphore:
     def __init__(self, value=1):
         self.value = value
-        self.cond = asyncio.Condition()
+        self.event_age = {}
+        self.queue = sortedcontainers.SortedSet(key=lambda event: self.event_age[event])
```

Comment:
Aged/sorted container seems like overkill, since ordering by age is just ... a list?  I think you can replace this with a deque (doubled-ended list that can push/pop either end).

Reply:
I wasn't 100% sure whether if I popped the first element of the list whether it would actually be the event I wanted, or other events could have been notified before I had the chance to delete the event. That's why I thought it was safer to do a remove operation for that specific event. remove is O(n) on a list, but in hindsight, maybe that's ok because it should be the first element in the list or close to it. I'll change it back to a list.